### ISSUE TITLE
[QST] Hungarian algorithm performance

### ISSUE THREAD

**Issue by bkj** [rapidsai/cugraph#1324]
Hi -- 

Are you able to share any benchmarking of the RAPIDS hungarian algorithm implementation that you've done?  

I'm comparing against the single-threaded CPU [lap](https://github.com/gatagat/lap) implementation, and RAPIDS seems quite a bit slower.  

On a 5000x5000 dense cost matrix, `lap` finishes in ~700ms, while `cugraph.hungarian` takes ~5s:
```python
import numpy as np
from time import perf_counter_ns

import cudf
import cugraph
from cugraph import hungarian

# --
# Data

n       = 5000
max_val = 1e6
np.random.seed(123)
C = np.random.uniform(0, max_val, size=(n, n)).round().astype(np.float32)

# --
# RAPIDS

df1        = cudf.DataFrame()
df1['src'] = cudf.Series(range(0, n, 1))
df1['key'] = 1

df2        = cudf.DataFrame()
df2['dst'] = cudf.Series(range(n, n + n, 1))
df2['key'] = 1

edges = df1.merge(df2, on='key')[['src', 'dst']]
edges = edges.sort_values(['src', 'dst']).reset_index()

edges['weight'] = C.flatten()

g = cugraph.Graph()
g.from_cudf_edgelist(edges, source='src', destination='dst', edge_attr='weight', renumber=False)

v1 = df1['src']

start_ns   = perf_counter_ns()
matching   = cugraph.hungarian(g, v1)
stop_ns    = perf_counter_ns()
rapids_elapsed_ms = (stop_ns - start_ns) / 1e6

cugraph_df = matching.merge(g.edgelist.edgelist_df,
                            left_on=['vertex', 'assignment'],
                            right_on=['src', 'dst'],
                            how='left')

cugraph_cost = cugraph_df['weights'].sum()

# --
# LAP

from lap import lapjv

start_ns = perf_counter_ns()
lap_cost, _, _ = lapjv(C)
stop_ns = perf_counter_ns()

lap_elapsed_ms = (stop_ns - start_ns) / 1e6
print(rapids_elapsed_ms, lap_elapsed_ms)

```

Any thoughts on what might be going on here?  Given the data structure that gets passed to `cugraph.hungarian`, perhaps your implementation is optimized for sparse assignment problems?

(I'm running on an AWS V100 using the RAPIDS docker image -- in case that's helpful.)

Thanks!

**Comment by bkj** [comment]
Ah OK cool -- I will run benchmarks for sparse applications as well.  I had thought that the Hungarian algorithm uses a dense representation, but perhaps I'm misremembering, or maybe this Hungarian implementation does something different.  

Thanks for your reply!

**Comment by afender** [comment]
The hungarian implementation actually uses a dense implementation.  We convert the sparse bipartite matrix into a dense matrix to do the computation. There is a C++ entry point in algorithms.hpp called cugraph::dense::hungarian which takes a dense representation. This is not currently exposed through the python layer because we are not really supporting dense inputs.

**Comment by afender** [comment]
We could do some detailed analysis, but alking to @ChuckHastings who added this to cuGraph, the cost could be dominated by:
- conversion of sparse to dense
- translation of the dense ids back to the sparse ids
- unrenumbering

**Comment by bkj** [comment]
Hmm OK -- interesting.  Yeah the conversion costs dominating seems like a likely culprit.  I'll look at the C++ entrypoint though -- I know `cugraph` isn't intended to support dense stuff, but AFAICT there aren't many GPU-based LAP solvers out there, so it'd be nice to have a Python interface to yours.

**Comment by ChuckHastings** [comment]
I have done some more detailed analysis, turning on some detailed timing within the C++ code.  My initial theory that the conversion costs might be the issue was incorrect, the conversion costs are a negligible amount of time.

The Jonker-Volgenant algorithm (the CPU algorithm you're calling called `lapjv`) is a different algorithm from the Hungarian algorithm that is implemented in cuGraph.  They solve the same problem (and generate the same answer), but the techniques are different and they have different performance characteristics.

I was able to reproduce your performance disparity above.  You are exercising this on a problem where the JV implementation is markedly faster than the Hungarian algorithm.

I ran some benchmarks of the Hungarian and JV algorithms across different problem sizes.  On first glance at the data, the Hungarian implementation will be more efficient when the max_val is smaller, the JV implementation will be more efficient when max_val is larger.  If you were to lower max_val to be 1000 (instead of 1000000), I believe you would find that Hungarian (on the GPU) would outperform the JV algorithm (on the CPU).  I don't know if this was an exploration for understanding or if the 1e6 value is representative of the actual problem you were trying to address.

While doing this analysis, I have created a python API to the dense version of the implementation.  You can track that in #1333.    This will simplify your above code dramatically, and will hopefully be included in version 0.18.

We will be looking at adding additional LAP algorithms into the library down the road.  Hopefully some of these new options will provide better performance in portions of the problem space where Hungarian does not.

**Comment by ChuckHastings** [comment]
Quick update.  The dense API from #1333 is merged, so it will be in version 0.18.  This should be available in the nightly packages if you want to experiment with that before 0.18 is released.
